File: python/125_Valid_Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if len(s) < 2:
            return True
        
        s = s.lower()
        left = 0
        right = len(s) - 1
        
        while left < right:
            while not s[left].isalnum():
                left += 1
                if left >= len(s):
                    return True
            
            while not s[right].isalnum():
                right -= 1
            
            if s[left] != s[right]:
                return False
            
            left += 1
            right -= 1
        
        return True
        

File: swift/125_valid_palindrome.swift
class Solution {
  func isPalindrome(_ s: String) -> Bool {
    let lowercasedS = s.lowercased()
    let arrayS = Array(lowercasedS)
    let onlyNumAlphaS: [Character] = removeNonAlphaNum(arrayS)

    return checkPalindrome(onlyNumAlphaS)
  }

  private func checkAlphaNum(_ char: Character) -> Bool {
    return char.isNumber || char.isLetter
  }

  private func removeNonAlphaNum(_ arrayS: [Character]) -> [Character] {
    var onlyAlphaNumChars: [Character] = []

    for char in arrayS {
      if checkAlphaNum(char) {
        onlyAlphaNumChars.append(char)
      }
    }

    return onlyAlphaNumChars
  }

  private func checkPalindrome(_ arrayS: [Character]) -> Bool {
    var left: Int = 0
    var right: Int = arrayS.count - 1

    while left < right {
      if arrayS[left] != arrayS[right] {
        return false
      }

      left += 1
      right -= 1
    }

    return true
  }
}
